### ISSUE TITLE
DyNet needs documentation!

### ISSUE THREAD

**Issue by neubig** [clab/dynet#85]
There is very little documentation and comments for dynet, except for the examples. It would be really nice to have this. Perhaps we can add comments to the code in doxygen format, etc. to facilitate automatic creation of documentation.

**Comment by neubig** [comment]
The chainer tutorial documents would be a good reference here:
http://docs.chainer.org/en/stable/tutorial/index.html

It looks like they used Sphinx.

**Comment by pmichel31415** [comment]
This looks good I'll try to set up sphinx and submit a PR

**Comment by neubig** [comment]
Awesome, that would be great.

**Comment by neubig** [comment]
Started a stem of some documentation here:
https://github.com/clab/dynet/blob/master/doc/index.md

I think one easy way to proceed is by commenting the code, then creating a script to automatically convert the comments into `.md` format for use in these files.

**Comment by pmichel31415** [comment]
Once #116 is merged, here's a TODO list : 

- [ ] Fill reference manual for all useful functions
 - [x] Optimizers
 - [ ] Operations
 - [ ] Builders
 - [ ] Minibatching
 - [ ] Multiprocessing
- [ ] Write contribution guidelines
 - [x] documentation guidelines
 - [ ] coding guidelines

**Comment by neubig** [comment]
I rearranged the operations and started documenting some of them, hopefully this will be done soon.

**Comment by neubig** [comment]
@pmichel31415: Do you think that we can use BibTeX in readthedocs? I took a look briefly but wasn't familiar enough with how readthedocs works to figure out how to do it.

**Comment by pmichel31415** [comment]
I'll take a look later this week

**Comment by pmichel31415** [comment]
So I've been looking around a little bit for references and it doesn't seem very practical in doxygen and/or sphinx (there's basically no canonical way to include bibtex references in sphinx).

I think the best solution is to keep adding references manually. In Doxygen you can add a link to the article (that's what I did in `training.h`) with `[article](link)`.

Or we could add a link to a "references" page in the sphinx doc (this would be a bit ugly but it would work).

**Comment by neubig** [comment]
Are we actually getting close to closing this? Of course there will always be some things that could use better documentation, but I think we've covered most of the major things, and anything remaining can be covered by individual issues. Thank you for the massive effort @pmichel31415

**Comment by pmichel31415** [comment]
Yes we can close this
